synthesizer/process/src/stack/mod.rs

```diff
@@ -468,3 +490,23 @@ impl<N: Network> PartialEq for Stack<N> {
 }
 
 impl<N: Network> Eq for Stack<N> {}
+
+// A helper enum to avoid cloning stacks.
+#[derive(Clone)]
+pub(crate) enum StackRef<'a, N: Network> {
```

Comment:
Hmm, I think we could just use `Arc<Stack<N>>` instead? Cloning an `Arc` is an atomic operation, and while it has some runtime penalty, it's a tiny one (not unlike using an `enum` as this one).

Reply:
In `cost.rs` and `get_number_of_calls` we wrap a `&Stack` in `StackRef::Internal` to avoid cloning.
We use the `StackRef` to persist an `&Stack` or `Arc<Stack>` across loop iterations.
Any ideas on how to avoid that pattern? If we have an alternative solution, then `StackRef` is no longer necessary.

Reply:
it seems to me that we could pass `&Arc<Stack>` to the cost functions and call `clone` on it when necessary, and it might be possible for `get_number_of_calls` to have a signature of
```
fn get_number_of_calls(self: &Arc<Self>, function_name: &Identifier<N>)
```

Reply:
Hmm, do you forsee any issues with a change like that? 
I feel like the `Arc` is less general than the reference.

Reply:
no, I'd only expect it to be simpler (i.e. no extra `StackRef` object); also, `Arc<T>` dereferences to `T`, so you can always obtain a simple reference from it

Reply:
I don't think this is always the case. `Stack::initialize` calls it without having an `Arc<Stack>`.
Unless you see any glaring issues, I think we could keep the interface as is.

Reply:
no, it's not a problem, it would only be a bit simpler (but also more restrictive)